src/OSPSuite.Core/Domain/Services/ParameterValuesCreator.cs

```diff
@@ -85,18 +89,72 @@ public ParameterValue CreateParameterValue(ObjectPath parameterPath, double valu
       }
 
       public IReadOnlyList<ParameterValue> CreateFrom(SpatialStructure spatialStructure, IReadOnlyList<MoleculeBuilder> molecules) =>
-         spatialStructure.PhysicalContainers.SelectMany(container => createFrom(container, molecules,
-            (x, c) => isLocalWithConstantFormula(x) && satisfiesContainerCriteria(x, c))).ToList();
+         spatialStructure.PhysicalContainers.SelectMany(container => createLocalFrom(container, molecules)).ToList();
 
-      public IReadOnlyList<ParameterValue> CreateExpressionFrom(IContainer physicalContainer, IReadOnlyList<MoleculeBuilder> molecules) =>
-         physicalContainer.GetAllContainersAndSelf<IContainer>(x => x.Mode.Is(ContainerMode.Physical)).SelectMany(container => createFrom(container, molecules,
-            (x, c) => x.IsExpression() && satisfiesContainerCriteria(x, c))).ToList();
+      public IReadOnlyList<ParameterValue> CreateExpressionFrom(IContainer physicalContainer, IReadOnlyList<MoleculeBuilder> molecules) => 
+         physicalContainer.GetAllContainersAndSelf<IContainer>(x => x.Mode.Is(ContainerMode.Physical)).SelectMany(container => createExpressionFrom(container, molecules)).ToList();
 
-      private IEnumerable<ParameterValue> createFrom(IContainer container, IReadOnlyList<MoleculeBuilder> molecules, Func<IParameter, IContainer, bool> createFor) =>
-         molecules.SelectMany(x => createFrom(container, x, createFor));
+      private IEnumerable<ParameterValue> createExpressionFrom(IContainer container, IReadOnlyList<MoleculeBuilder> molecules) => molecules.SelectMany(x => createExpressionFrom(container, x));
 
-      private IEnumerable<ParameterValue> createFrom(IContainer container, MoleculeBuilder molecule, Func<IParameter, IContainer, bool> createFor) =>
-         molecule.Parameters.Where(x => createFor(x, container)).Select(x => CreateParameterValue(objectPathForParameterInContainer(container, x.Name, molecule.Name), x));
+      private void initializeFormulasFor(IReadOnlyList<ParameterValue> parameterValues, IReadOnlyList<ExpressionParameter> expressionParameters)
+      {
+         var formulaCache = new FormulaCache();
+         parameterValues.Each(formulaTarget =>
+         {
+            var formulaSource = formulaSourceFor(expressionParameters.AllByName(formulaTarget.Name).ToList(), formulaTarget);
+
+            if (formulaSource != null)
+               formulaTarget.Formula = _cloneManager.Clone(formulaSource.Formula, formulaCache);
+         });
+      }
+
+      private bool shouldCloneFormulaFor(ParameterValue parameterValue) => parameterValue.Formula == null || parameterValue.Formula.IsConstant();
```

Comment:
I don't quite understand why the function returns `true` if `parameterValue.Formula = null`

Reply:
Oh, that's because I don't want to overwrite it if the ParameterValue already has a formula due to the molecule parameter it's based on.

Reply:
I added a comment where this is used.